packages/amplify-cli-core/src/utils/packageManager.ts

```diff
@@ -43,9 +44,31 @@ class YarnPackageManager implements PackageManager {
   version?: SemVer;
 
   getRunScriptArgs = (scriptName: string) => [scriptName];
-  getInstallArgs = (buildType = BuildType.PROD) => {
+  getInstallArgs = (buildType = BuildType.PROD, resourceDir = '') => {
     const useYarnModern = this.version?.major && this.version?.major > 1;
-    return (useYarnModern ? ['install'] : ['--no-bin-links']).concat(buildType === 'PROD' ? ['--production'] : []);
+    /**
+     * Since Yarn 2, resourceDir needs to be treated as a separate project,
+     * otherwise it'll be hoisted to use the lock.file from the parent directory,
+     * so we need to create a lock file for it.
+     * ref: https://github.com/yarnpkg/yarn/issues/5716#issuecomment-817330338
+     */
+    if (useYarnModern) {
+      if (fs.existsSync(`${resourceDir}/${this.lockFile}`)) {
+        console.log(`${resourceDir}/${this.lockFile} already exists.`);
+      } else {
+        fs.writeFile(`${resourceDir}/${this.lockFile}`, '', (err) => {
+          if (err) {
+            console.error(`Error creating ${resourceDir}/${this.lockFile}`, err);
+          } else {
+            console.log(`${resourceDir}/${this.lockFile} created successfully.`);
+          }
+        });
```

Comment:
Shouldn't we be using sync methods or promisified ones instead of callbacks? `execa.sync` after will execute right away whether file was written or not.

Reply:
Done! 1df95f1d4136e841a07181ab13e2b454421c4030